src/uct/cuda/cuda_copy/cuda_copy_md.c

```diff
@@ -202,14 +217,107 @@ UCS_PROFILE_FUNC(ucs_status_t, uct_cuda_copy_mem_dereg,
     return UCS_OK;
 }
 
-static ucs_status_t uct_cuda_copy_mem_alloc(uct_md_h md, size_t *length_p,
+static ucs_status_t
+uct_cuda_copy_mem_alloc_fabric(uct_cuda_copy_alloc_handle_t *alloc_handle)
+{
+#if HAVE_CUDA_FABRIC
+    CUdevice_attribute attrib   = CU_DEVICE_ATTRIBUTE_HANDLE_TYPE_FABRIC_SUPPORTED;
+    size_t granularity          = 0;
+    CUmemAllocationProp prop    = {};
+    CUmemAccessDesc access_desc = {};
+    int fabric_supported;
+    ucs_status_t status;
+    CUdevice cu_device;
+
+    status = UCT_CUDADRV_FUNC_LOG_ERR(cuCtxGetDevice(&cu_device));
+    if (status != UCS_OK) {
+        return status;
+    }
+
+    status = UCT_CUDADRV_FUNC_LOG_ERR(cuDeviceGetAttribute(&fabric_supported,
+                                      attrib, cu_device));
+    if (status != UCS_OK) {
+        return status;
+    }
+
+    prop.type                 = CU_MEM_ALLOCATION_TYPE_PINNED;
+    prop.requestedHandleTypes = CU_MEM_HANDLE_TYPE_FABRIC;
+    prop.location.type        = CU_MEM_LOCATION_TYPE_DEVICE;
+    prop.location.id          = cu_device;
+
+    status = UCT_CUDADRV_FUNC_LOG_ERR(cuMemGetAllocationGranularity(&granularity, &prop,
```

Comment:
1. cache this on MD
2. there are comments provided to the original PR - can you pls review and apply them?